Add reexport option and `import ... as ...` flavour

dartt0n opened the pull request:
Reexport options allows to generate imports in format
```
import <module> as <module_name>
from <module> import <object> as <object>
```

dartt0n committed:
Add reexport option

Reexport options allows to generate imports in format ``` import
<module> as <module_name> from <module> import <object> as <object> ```

Erotemic commented:
Logic looks good. We will need a test to make sure the behavior doesn't get broken by accident and a CHANGELOG note too.

dartt0n committed:
Add tests for reexport feature

dartt0n committed:
Bump up version and update CHANGELOG.md

dartt0n commented:
@Erotemic I have added tests and updated the version and changelog. I'm not sure if I did the last part correctly, so I'll be glad to receive feedback and I'm ready to correct the code.

Erotemic commented:
I made a few changes to the PR to add some documentation and do some minor refactoring. I don't seem to have permission to push to your branch, could you either give that to me, or fetch the branches from this repo and then cherry pick 0334fe088d1931b3cba63d978608396d76414156 onto your branch.

I also see that when I specify `--reexport` it still generates an `__all__` variable, which I don't think is necessary, right? Isn't that the point of this to avoid needing that?

I also have reservations about using `--flag` style variables. The issue is that unless you do a lot of extra stuff with argparse there is no way to get them into a key/value format (e.g. `--flag=True`). That means if want to default the `reexport` option to true, then we have to keep an unused CLI argument and then add a `--no-rexport` option, which just gets cumbersome. This is one of my main motivations for writing https://github.com/Kitware/scriptconfig (although I want mkinit to only rely on the stdlib, so we can't use that here).

However, there might be a better option. We can use a "modal" or "categorical" CLI argument. We could have the new CLI option be `--export_style` which should currently default to "all", or you could enable reexports by changing it to `--export_style=as` or something like that. That way if someone hard codes `--export_style=as` in their command it wouldn't break if we decided to default to it in the future. 

I'm not 100% sure on what the name of the argument and its choices should be. I think I'd like something like this:

```python
parser.add_argument(
    "--export-style",
    "--export_style",
    dest="export_style",
    choices=["all", "as"],
    default="all",
    help=(
        "Style used to denote exports. "
        "'all' uses __all__; "
        "'as' uses 'import ... as ...' form."
    ),
)
```

open to feedback on this idea.

EDIT:

This doesn't need to be solved in this PR, but I'm also seeing the option `--norespect_all` which might be better named as `--norespect_exports` (although we can't do that for backwards compat reasons, at least without an alias), but maybe that option should be sensitive to `import X as X` exports? Or "as-style" exports might be a better way to refer to it?

MorganStair commented:
A few quick things I noticed, in order of priority:

1. BUG: In mkinit/formatting.py (the exposed_submodules bit), submodule imports get appended once unconditionally and then appended AGAIN (either reexport style or import style). That will duplicate lines / emit both forms when --reexport is on.


2. BUG(?): _make_reexports_str() has a branch that can generate import {name} as {name}. If {name} ever ends up dotted (e.g., pkg.submod), that’s invalid Python (as alias can’t contain .). Might be worth confirming that branch can’t happen, or adjusting it to use a valid alias (last segment) / a from pkg import submod as submod style.

3. Tests currently check “contains these lines,” which wouldn’t catch duplicates. A quick “exact output” / “count == 1” assertion for the submodule lines would make that solid.

4. Codecov still shows some uncovered lines (__main__.py + a couple in formatting.py). A tiny CLI-level test for --reexport would probably close most of that.

dartt0n commented:
Thanks for the detailed feedback! I will try to come back to the solution during the working week. 
Initially, I tried to make as few changes as possible so as not to accidentally affect the existing solution. 

I don't have much experience with supporting CLI solutions, so this feedback is very insightful. I will pay attention to the name of the flags and their future support, as well as introduce more strict tests to make sure no duplication or invalid code generation happens.
